Change sharing map so that leafs can be directly attached to internal nodes

danpoe opened the pull request:
Previously all leafs were stored in a linked list. However, this was unnecessary as collisions can occur only between nodes at the bottom of the tree. Now only leafs at the bottom are stored in a linked list, and all other leafs are direct successors of internal nodes. 

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- n/a Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- n/a The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

danpoe committed:
Unify leaf nodes with internal nodes and container nodes

Previously the leaf nodes were a separate type, and internal nodes and container
nodes were represented by the same type. This changes the representation such
that all nodes are represented by sharing_nodet. This will allow an internal
node to directly point to a leaf node (instead of having to point to a container
node which in turn points to a leaf node).

danpoe committed:
Adapt methods for finding, erasing, and inserting elements into the sharing map

This adapts the methods for the case where a leaf can directly be attached to
internal nodes.

danpoe committed:
Adapt iteration over the sharing map

Adapt get_view() and get_delta_view() for the case where a leaf node is directly
attached to an internal node.

danpoe committed:
Remove obsolete method is_singular() in sharing_mapt

danpoe committed:
Use emplace() in sharing_nodet

danpoe committed:
Return reference from methods that cannot return a nullpointer

Previously some methods returned a pointer but would never return nullptr. This
changes their return type to references instead.

danpoe committed:
Use type nodet to represent sharing map nodes

Previously there existed a separate leaft type. Now all nodes are of the type
sharing_nodet<...> (of which nodet is an lias). This replaces all node types by
nodet.

danpoe committed:
Reorder types in sharing_nodet to optimize for common case

When destructt::destruct() of a small shared n-way pointer is called, it first
determines the type of its pointee. For example, if the parameter pack Ts...
is instantiated to T1, T2, T3, then it first checks if the pointer points to T3,
then to T2, etc. Thus, the least checks are performed if it is most likely that
the pointer points to an object of type T3, then T2, etc.

danpoe committed:
Add new unit tests which check the shape of the map and adapt existing ones

The shape of a sharing map is now different than before, as now leafs can be
directly attached to internal nodes. This adapts the various unit tests that
check that the map has a certain shape.

danpoe commented:
On a set of methods from Apache Tika, the sum of the runtimes of the sharing map methods (timing just those) improved by 28%. Overall symex was 7% faster.

smowton commented on src/util/sharing_map.h:
```diff
@@ -193,10 +193,8 @@ class sharing_mapt
-  typedef sharing_node_leaft<key_type, mapped_type> leaft;
-
-  typedef sharing_node_baset baset;
+  typedef sharing_nodet<key_type, mapped_type> innert;
```
(random line) commit message: where -> were :p

smowton commented on src/util/sharing_node.h:
```diff
@@ -70,14 +69,14 @@ const T *as_const(T *t)
 
 // Inner nodes (internal nodes or container nodes)
```
Comment out of date now there are 3 possibilities?

smowton commented on src/util/sharing_map.h:
```diff
@@ -518,20 +518,8 @@ class sharing_mapt
-
-    return lp;
-  }
+  leaft *get_leaf_node(const key_type &k);
```
(random line) commit message: leafs -> leaf

smowton commented on src/util/sharing_map.h:
```diff
@@ -1077,10 +1079,21 @@ SHARING_MAPT2(const, innert *)::get_container_node(const key_type &k) const
+    }
 
-    key >>= chunk;
+    if(ip->is_leaf())
```
These are mutually exclusive, right? If so use `else if` to clarify that.

smowton commented on src/util/sharing_map.h:
```diff
@@ -1111,62 +1124,72 @@ SHARING_MAPT(void)::erase(const key_type &k)
+      continue;
+    }
+
+    if(ip->is_leaf())
```
`else if`

smowton commented on src/util/sharing_map.h:
```diff
@@ -1111,62 +1124,72 @@ SHARING_MAPT(void)::erase(const key_type &k)
+      return;
+    }
 
     if(ip->is_container())
```
`else if`

smowton reviewed (approved):
LGTM (at moderate depth, haven't walked through the algorithms in detail to make sure there are no missing cases). Surprised not to see any updates to the top-level docs for the data structure though?

danpoe committed:
Update high-level sharing map documentation

Update the documentation to reflect the fact that the sharing map is now a
variable-height tree.

tautschnig commented on src/util/sharing_map.h:
```diff
@@ -1233,88 +1299,93 @@ ::insert(const key_type &k, valueU &&m)
 
   INVARIANT(
-    !value_equalt()(as_const(lp)->get_value(), m),
+    !value_equalt()(as_const(lp).get_value(), m),
```
Visual Studio complains that the use of `as_const` is ambiguous here. You might have add/use a static cast here.

danpoe committed:
Update best case complexity of sharing map operations

Since the sharing map is now a variable-height tree, the best case complexity of
various operations has changed.

allredj reviewed (commented):
:warning:
This PR failed Diffblue compatibility checks (cbmc commit: e7a5cf1d7dd53ebc51072e994b92f60d5ca96cf1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112914521
Status will be re-evaluated on next push.
Common spurious failures include: the cbmc commit has disappeared in the mean time (e.g. in a force-push); the author is not in the list of contributors (e.g. first-time contributors); compatibility was already broken by an earlier merge.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 872f2030df952458fd0472fc25e4c6ba685dc1b1).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112923269

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: d43834165ffef151630e3a5246039ef1463a9a54).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/112940281
